src/liballoc/sync.rs

```diff
@@ -751,6 +751,77 @@ impl<T: ?Sized> Arc<T> {
         this.inner().strong.load(SeqCst)
     }
 
+    /// Increments the strong reference count on the `Arc<T>` associated with the
+    /// provided pointer by one.
+    ///
+    /// # Safety
+    ///
+    /// The pointer must have been obtained through `Arc::into_raw`, and the
+    /// associated `Arc` instance must be valid (i.e. the strong count must be at
+    /// least 1) for the duration of this method.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(arc_mutate_strong_count)]
+    ///
+    /// use std::sync::Arc;
+    ///
+    /// let five = Arc::new(5);
+    ///
+    /// unsafe {
+    ///     let ptr = Arc::into_raw(five);
+    ///     Arc::incr_strong_count(ptr);
+    ///
+    ///     // This assertion is deterministic because we haven't shared
+    ///     // the `Arc` between threads.
+    ///     let five = Arc::from_raw(ptr);
+    ///     assert_eq!(2, Arc::strong_count(&five));
+    /// }
+    /// ```
+    #[inline]
+    #[unstable(feature = "arc_mutate_strong_count", issue = "none")]
+    pub unsafe fn incr_strong_count(ptr: *const T) {
+        // Retain Arc, but don't touch refcount by wrapping in ManuallyDrop
+        let arc = mem::ManuallyDrop::new(Arc::<T>::from_raw(ptr));
+        // Now increase refcount, but don't drop new refcount either
+        let _arc_clone: mem::ManuallyDrop<_> = arc.clone();
+    }
+
+    /// Decrements the strong reference count on the `Arc<T>` associated with the
+    /// provided pointer by one.
+    ///
+    /// # Safety
+    ///
+    /// The pointer must have been obtained through `Arc::into_raw`, and the
+    /// associated `Arc` instance must be valid (i.e. the strong count must be at
+    /// least 1) for the duration of this method.
```

Comment:
Hm, okay, we should update this -- it currently indicates that it is unsafe to use this method to release the final Arc (and free the backing storage), which is not true.

I'm not quite sure how to phrase it, though.

Reply:
Ah good point; perhaps this could work:
```suggestion
    /// least 1) when invoking this method. This method can be used to release the final
    /// `Arc` and backing storage, but should not be called after the final `Arc` has been
    /// released.
```

Reply:
*after* there is not quite right -- we can't race with that release either -- but I'm struggling with wording that isn't super long for that... maybe this is fine as is?

Reply:
Yeah I think it's fine as-is since, especially since we're not yet seeking to stabilize this yet. I think people will understand what the constraints here are.

Though we should probably mention it as something that could be improved on in the tracking issue.